appsec-kit-v7/src/main/java/com/vaadin/appsec/v7/service/NotificationInitializer.java

```diff
@@ -37,91 +36,49 @@ public class NotificationInitializer {
 
-    /**
-     * Initializes the notification service
-     *
-     * @param vaadinService
-     *            current VaadinService instance
-     */
-    static void serviceInit(VaadinService vaadinService) {
-        if (isDebugMode(vaadinService)) {
-            vaadinService.addSessionInitListener(e -> {
-                e.getSession().addUIProvider(new AppSecUIProvider());
-                startNotificationTask(e.getSession());
-            });
-            LOGGER.info("NotificationInitListener initialized.");
+    void serviceInit(VaadinService service) {
+        if (isDebugMode(service)) {
+            service.addSessionInitListener(this::subscribeSessionToScanEvents);
+            service.addSessionDestroyListener(this::removeSessionRegistration);
+            LOGGER.info("Subscribed to AppSec Kit scan events");
         }
     }
 
-    private static void startNotificationTask(final VaadinSession session) {
-        final String sessionId = session.getSession().getId();
-        final ScheduledFuture<?> scheduledNotificationCheck = AppSecService
-                .getInstance().getConfiguration().getTaskExecutor()
-                .scheduleAtFixedRate(() -> {
-                    if (isSessionOpen(session)) {
-                        session.access(() -> {
-                            session.getUIs().forEach(
-                                    ui -> notifyUiIfNeeded(session, ui));
-                        });
-                    } else {
-                        ScheduledFuture<?> thisCheck = scheduledChecks
-                                .get(sessionId);
-                        if (thisCheck != null) {
-                            thisCheck.cancel(true);
-                            scheduledChecks.remove(sessionId);
-                        }
-                    }
-                }, NOTIFICATION_CHECK_INTERVAL, NOTIFICATION_CHECK_INTERVAL,
-                        TimeUnit.MILLISECONDS);
-        scheduledChecks.put(sessionId, scheduledNotificationCheck);
-        LOGGER.info(
-                "NotificationInitListener notification thread initialized.");
+    private void removeSessionRegistration(SessionDestroyEvent e) {
+        VaadinSession session = e.getSession();
+        Registration registration = scanEventRegistrations.get(session);
+        if (registration != null) {
+            registration.remove();
+        }
+    }
 
+    private void subscribeSessionToScanEvents(SessionInitEvent e) {
+        VaadinSession session = e.getSession();
+        session.addUIProvider(new AppSecUIProvider());
+        AppSecService appSecService = AppSecService.getInstance();
+        Registration scanEventRegistration = appSecService
+                .addScanEventListener(scanEvent -> {
+                    int newVulns = scanEvent.getNewVulnerabilities().size();
+                    if (isSessionOpen(session) && newVulns > 0) {
+                        session.getUIs().forEach(this::doNotifyUI);
```

Comment:
With `session.access(...)` function the notification popup works as expected.
```suggestion
session.access(() -> session.getUIs().forEach(
        this::doNotifyUI));
```

Reply:
Now this is done per-UI in `doNotifyUI`.